### ISSUE TITLE
RFC:  Addition of MIT 6.S191

### ISSUE THREAD

**Issue by aayushsinha0706** [ossu/computer-science#1013]
**Problem:**
MIT has its course on deep learning MIT 6.S191 that is updated every year and is a high quality resource attented by many students around the globe. 

**Duration:**
26 April 2022

**Background:**
MIT 6.S191 Intro to deep learning is a high qualtiy resource by Alexander Amini sponsored by Google and many other tech giants. The course doesnt require much pre-requisite knowledge. MIT 6.S191 states that "We are expecting very elementary knowledge of linear algebra and calculus. How to multiply matrices, take derivatives and apply the chain rule. Familiarity in Python is a big plus as well. The course will be beginner friendly since we have many registered students from outside of computer science."

**Proposal:**
Not quite sure wether to add this course on curriculum as Advanced Applications or in extra/courses. I will leave it to the OSSU CS team to decide,

Add [MIT 6.S191 Introduction to Deep Learning](http://introtodeeplearning.com) as Advanced Applications or in extra/courses.

**Comment by waciumawanjohi** [comment]
I think such a course would best fit into an Advanced AI section. We can see the relevant section of the CS2013 here:
https://www.acm.org/binaries/content/assets/education/cs2013_web_final.pdf#page=124

Neural networks are elective topics, and so most appropriate for the advanced section of the curriculum. I'll also note that currently the single Machine Learning course is in Advanced Applications, while there are AI (or IS, Intelligent System) topics that are Tier 2 in CS2013. That would suggest that some resource should be in the core offering.

**Comment by spamegg1** [comment]
<details>
<summary> (Click to expand) @waciumawanjohi without detracting too much from the current issue...</summary>

> I'll also note that currently the single Machine Learning course is in Advanced Applications...

There is Machine Learning in Core Applications, no? Is that considered a different topic by the guidelines? Excuse me if I'm ignorant. 

> Neural networks are elective topics, and so most appropriate for the advanced section of the curriculum. 

The ML course in Core Apps touches upon Neural Networks for two Lectures, with one Programming Assignment for implementing it. Hah! The world cares not for our nice neat distinctions.

> while there are AI (or IS, Intelligent System) topics that are Tier 2 in CS2013. That would suggest that some resource should be in the core offering.

Ah, I guess there is some semantic distinction between AI and ML? Looking over the relevant section (IS) in the guidelines, everything is listed: AI, Robotics, ML. Only Basic ML is under Tier 2, with the other two under electives. I suppose Deep Learning would fall under "Advanced ML" although it's not mentioned explicitly.

![Screenshot from 2022-05-15 23-05-56](https://user-images.githubusercontent.com/4255997/168491836-a9234a18-0fea-4a4d-9c78-d49e0ba7ab4e.png)

To complicate things even further, Deep Learning has sections on self-driving, which is considered (probabilistic) Robotics... and both DL and ML courses have sections on Computer Vision (another elective). What a nice big semantic mess. I don't think we can disentangle it and neatly separate things.

However IS/Fundamental Issues is not covered in Core. Only a very small part of IS/Basic Search Strategies is covered in Core Theory. (IS/Basic Knowledge Representation and Reasoning is covered by Math for CS.) So I see your point there. 

Let me attempt to formulate an opinion. I think Core is fine as it is. The little bits that fall through the cracks will have to be covered in Advanced, until some better alternative comes along.

</details>

> I think such a course would best fit into an Advanced AI section.

The idea of creating an Advanced AI section is an interesting one. If I understand correctly, learners would choose that path right after finishing Core, so it would become an additional alternative to Advanced Programming/Advanced Theory/Advanced Systems?

I'd be fine with 
- adding MIT Deep L to Advanced Apps, or 
- offering a choice "pick one of..." between it and Robotics, or 
- creating a new Advanced AI section with both courses under it. (This is probably the best one.)

**Comment by waciumawanjohi** [comment]
> There is Machine Learning in Core Applications, no?

You are correct, I was mistaken when I said the Machine Learning course was in Advanced Applications. There is in fact no Advanced Applications section and the ML course is in Core Applications.

> some semantic distinction between AI and ML?

Yes. As [this article](https://www.ibm.com/cloud/blog/ai-vs-machine-learning-vs-deep-learning-vs-neural-networks) points out, Deep Learning is a subfield of Machine Learning, which is a subfield of AI. An example of a field in AI that is not Machine Learning would be minimax, a version of adversarial search that can be used to create game playing agents (e.g. an agent that would compete in checkers). An example of a subfield of Machine Learning that is not Deep Learning would be decision trees.

> learners would choose that path right after finishing Core, so it would become an additional alternative to Advanced Programming/Advanced Theory/Advanced Systems?

Exactly. I agree with you, that section should have a Deep Learning course as well as a robotics course. A good NLP (natural language processing) and Computer Vision course would be appropriate. All the top hits from the chart you shared.

**Comment by waciumawanjohi** [comment]
For a CV offering, I think this looks the most appropriate:
https://www.coursera.org/learn/introduction-computer-vision-watson-opencv

Touches on deep learning, but not every method is deep learning.

[Here's another option](https://www.coursera.org/learn/computer-vision-basics). My concern is that this doesn't look to get into actual CV methods, focusing instead on background knowledge.

The CV subreddit maintains a list of resources in [their wiki](https://old.reddit.com/r/computervision/wiki/index).

**Comment by waciumawanjohi** [comment]
There is [a similar wiki from the robotics subreddit.](https://www.reddit.com/r/robotics/wiki/resources)

**Comment by aayushsinha0706** [comment]
> Exactly. I agree with you, that section should have a Deep Learning course as well as a robotics course. A good NLP (natural language processing) and Computer Vision course would be appropriate. All the top hits from the chart you shared.

Few courses that I would love to recommend for Advanced AI 

[Stanford CS224n NLP with Deep Learning](https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1214/)

>Natural language processing (NLP) or computational linguistics is one of the most important technologies of the information age. Applications of NLP are everywhere because people communicate almost everything in language: web search, advertising, emails, customer service, language translation, virtual agents, medical reports, politics, etc. In the last decade, deep learning (or neural network) approaches have obtained very high performance across many different NLP tasks, using single end-to-end neural models that do not require traditional, task-specific feature engineering.

> In this course, students will gain a thorough introduction to cutting-edge research in Deep Learning for NLP. Through lectures, assignments and a final project, students will learn the necessary skills to design, implement, and understand their own neural network models, using the [Pytorch](https://pytorch.org/) framework.


[Deepmind X UCL Reinforcement Learning](https://www.deepmind.com/learning-resources/reinforcement-learning-lecture-series-2021)

> Reinforcement learning is an area of machine learning concerned with how intelligent agents ought to take actions in an environment in order to maximize the notion of cumulative reward. Reinforcement learning is one of three basic machine learning paradigms, alongside supervised learning and unsupervised learning.                               - Wikipedia

CS 2013 also recommends reinforcement learning for learning of advanced machine learning

• Reinforcement learning
o Exploration vs. exploitation trade-off 
o Markov decision processes
o Value and policy iteration

### Advanced AI

Courses | Duration | Effort | Prerequisites
:-- | :--: | :--: | :--:
[Introduction to Deep Learning](http://introtodeeplearning.com) | 12 weeks | 8-12 hours/week | Familiarity with Python ,Linear algebra and calculus 
Natural Language Processing with Deep Learning [(video lectures)](https://www.youtube.com/playlist?list=PLoROMvodv4rOSH4v6133s9LFPRHjEmbmJ) [(course website and assignments)](https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1214/) | 10 weeks | 8-12 hours/week | Proficiency in Python (NumPy and PyTorch) , Calculus, Linear Algebra,Probability & Statistics and Machine Learning
[Introduction to Computer Vision and Image Processing](https://www.coursera.org/learn/introduction-computer-vision-watson-opencv) | 6 weeks | 4-5 hours/week | Proficiency in Python 
[Reinforcement Learning](https://www.deepmind.com/learning-resources/reinforcement-learning-lecture-series-2021) | 12 weeks | 6-8 hours/week |Proficiency in Python (Must be comfortable converting algorithms and pseudocode into Python), Calculus, Linear Algebra, Probability & Statistics and Machine Learning
[Modern Robotics (specialization) ](https://www.coursera.org/specializations/modernrobotics) | 13 weeks | 5-10 hours/week | Freshman Physics, linear algebra, calculus, [linear order differential equations](https://www.khanacademy.org/math/differential-equations)

**Comment by spamegg1** [comment]
I took [Artificial Intelligence for Robotics](https://www.udacity.com/course/artificial-intelligence-for-robotics--cs373) on Udacity years ago. 2012 maybe? Great course! It's one of those unmaintained free Udacity courses. It still uses Python 2. I tried a few assignments and the grader seems to be working, so it should be doable.

**Comment by waciumawanjohi** [comment]
In considering an Advanced AI track, we should look to introduce learners at the undergraduate level to the breadth of the topic. The classic textbook on AI is [Artificial Intelligence: A Modern Approach](https://en.wikipedia.org/wiki/Artificial_Intelligence:_A_Modern_Approach). This text is broken into 5 parts:

## Problem-solving    
Solving Problems by Searching    
Search in Complex Environments    
Adversarial Search and Games    
Constraint Satisfaction Problems
## Knowledge, reasoning, and planning    
Logical Agents    
First-Order Logic    
Inference in First-Order Logic    
Knowledge Representation    
Automated Planning
## Uncertain knowledge and reasoning    
Quantifying Uncertainty    
Probabilistic Reasoning    
Probabilistic Reasoning over Time    
Probabilistic Programming    
Making Simple Decisions    
Making Complex Decisions    
Multiagent Decision Making ... 
## V Machine Learning    
Learning from Examples    
Learning Probabilistic Models    
Deep Learning    
Reinforcement Learning
## Communicating, perceiving, and acting    
Natural Language Processing    
Deep Learning for Natural Language Processing    
Computer Vision    
Robotics

The courses listed so far concentrate on two of these areas: Learning and Perception/Action.

**Comment by waciumawanjohi** [comment]
I ran across this trio of courses on optimization:
[Basic Modeling for Discrete Optimization](https://www.coursera.org/learn/basic-modeling)
[Advanced Modeling for Discrete Optimization](https://www.coursera.org/learn/advanced-modeling)
[Solving Algorithms for Discrete Optimization](https://www.coursera.org/learn/solving-algorithms-discrete-optimization)

Interesting possibilities

**Comment by aayushsinha0706** [comment]
Three courses that can be discussed upon for Advanced Artificial Intelligence

[Artificial Intelligence: Search Methods for Problem Solving](https://nptel.ac.in/courses/106106126)
[Artificial Intelligence: Knowledge Representation and Reasoning](https://nptel.ac.in/courses/106106140)

Both the courses are offered by Indian Institute of Technology Madras and covers Problem-solving and Knowledge, reasoning, and planning in depth

the third one is from MITx
[Computational Probability and Inference](https://www.edx.org/course/computational-probability-and-inference)

the above course is the only course that discuss probabilistic inference in detail.

**Comment by nicol4us** [comment]
I am doing Java programming from University of Helsinki and I got amazed how good their course got organized. So I explored their courses and found that they also have AI course.  So probably you can try to check if it meets with our needs. Please find below the link.
[Elements of AI](https://www.elementsofai.com/)

**Comment by aayushsinha0706** [comment]
I happen to enounter a really good course on AI research. Its Harvard CS197 AI research experiences and Harvard has made its lecture notes free. I propose to add this in extra-readings or extra-courses. https://www.cs197.seas.harvard.edu

**Comment by aayushsinha0706** [comment]
At the time this RFC was opened it was just about including deep learning course of MIT in extras. It quickly became the RFC for advanced AI. There are numerous high quality resources on AI by universities but unfortunately we cannot include them all. Although I hope some of them can be included in extras.

Few concerns OSSU lacks some core studies in AI  like 
1. IS/Basic Knowledge Representation and Reasoning 
2. IS/Basic Search Strategies(some of this is covered in algorithms)

I recommend this course by UC Berkely Summer 2016 to cover above studies
[Introduction to Artificial Intelligence](https://inst.eecs.berkeley.edu/~cs188/su16). It redirects to an edx edge platform which is different from original edx. But the course is not behind a paywall and grader might work fine.

But do I recommend in core applications? No, I guess Andrew ng machine learning is more than enough exposure to AI for everyone who is following curriculum.

But with the battle of pre-requisites. Every now and then we see comments in discord like why linear algebra is pre-requisite of machine learning when it is in advanced math. Intro to AI requires background in mathematics which belongs to advanced math and hence it will make more sense to for it to be the first course in Advanced AI

Now what areas CS2013 recommend for an advanced AI

IS/Advanced Search
IS/Advanced Representation and Reasoning
IS/Reasoning Under Uncertainty
IS/Agents
IS/Natural Language Processing
IS/Advanced Machine Learning
IS/Robotics

**Comment by Choubs01** [comment]
What are the conditions for adding a field in advanced topics as opposed to a specialization? Me personally I was planning on doing an AI specialization (https://www.coursera.org/specializations/machine-learning-introduction?#courses) once I was done with advanced section.

**Comment by aayushsinha0706** [comment]
> What are the conditions for adding a field in advanced topics as opposed to a specialization? Me personally I was planning on doing an AI specialization (https://www.coursera.org/specializations/machine-learning-introduction?#courses) once I was done with advanced section.

Advanced topics are more like graduate level class or courses that are mentioned to taken as an elective in CS2013. Also for Andrew Ng ML Class above we are recommending it to be in core. Check this RFC #1118

**Comment by Dan-Kot** [comment]
The advanced AI outline posted in this outline looks good

> ### Advanced AI
> Courses	Duration	Effort	Prerequisites
> [Introduction to Deep Learning](http://introtodeeplearning.com)	12 weeks	8-12 hours/week	Familiarity with Python ,Linear algebra and calculus
> Natural Language Processing with Deep Learning [(video lectures)](https://www.youtube.com/playlist?list=PLoROMvodv4rOSH4v6133s9LFPRHjEmbmJ) [(course website and assignments)](https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1214/)	10 weeks	8-12 hours/week	Proficiency in Python (NumPy and PyTorch) , Calculus, Linear Algebra,Probability & Statistics and Machine Learning
> [Introduction to Computer Vision and Image Processing](https://www.coursera.org/learn/introduction-computer-vision-watson-opencv)	6 weeks	4-5 hours/week	Proficiency in Python
> [Reinforcement Learning](https://www.deepmind.com/learning-resources/reinforcement-learning-lecture-series-2021)	12 weeks	6-8 hours/week	Proficiency in Python (Must be comfortable converting algorithms and pseudocode into Python), Calculus, Linear Algebra, Probability & Statistics and Machine Learning
> [Modern Robotics (specialization) ](https://www.coursera.org/specializations/modernrobotics)	13 weeks	5-10 hours/week	Freshman Physics, linear algebra, calculus, [linear order differential equations](https://www.khanacademy.org/math/differential-equations)

**Comment by Ai80-8** [comment]
Delete it all
